packages/test-sync/src/androidAndroidTest/kotlin/io/realm/kotlin/test/mongodb/shared/SyncSessionTests.kt

```diff
@@ -570,6 +573,34 @@ class SyncSessionTests {
         }
     }
 
+    @Test
+    fun connectionState() = runBlocking {
+        Realm.open(createSyncConfig(user)).use { realm: Realm ->
+            // We don't know what the state will be, but just verify that we can retrieve it
+            // without issues
+            assertNotNull(realm.syncSession.connectionState)
+        }
+    }
+
+    @Test
+    fun connectionState_asFlow() = runBlocking {
```

Comment:
There seems to be some tests in Java that indicate that removing the connection listener will return DISCONNECTED. Should we add a `cancel` test for the Flow in Kotlin?

```
    @Test
    fun removeConnectionListener() = looperThread.runBlocking {
        Realm.getInstance(syncConfiguration).use { realm ->
            val session: SyncSession = realm.syncSession
            val listener1 = ConnectionListener { oldState: ConnectionState?, newState: ConnectionState ->
                if (newState == ConnectionState.DISCONNECTED) {
                    fail("Listener should have been removed")
                }
            }
            val listener2 = ConnectionListener { oldState, newState ->
                if (newState == ConnectionState.DISCONNECTED) {
                    looperThread.testComplete()
                }
            }
            session.addConnectionChangeListener(listener1)
            session.addConnectionChangeListener(listener2)
            session.removeConnectionChangeListener(listener1)
        }
    }
```

Reply:
The above test verifies that removing the listener is in fact removing it, as it is not failing on a disconnect on close. So there are two things in play. One is canceling the flow, and the other is that realm-java actually receives a disconnect when the realm is closed. I don't see such an event in my current implementation 👀  but should we keep the flow open and await a disconnected event or would we actually rather complete the flow on `close` ... as we have done a number of other places 🤔 

I guess this is a general behavioral pattern that we should then also consider for progress listeners.

Reply:
Hmm, I think we do need some well-defined semantics for the `close` case. I think we should complete the flow, but whether or not we should emit a DISCONNECT before doing it...it would probably make the most sense, even though it probably ends up making the implementation a bit more complex.

But yeah, we should probably also close the progress listeners.

Reply:
I have added infrastructure to close flows when the realm is closed and have added test for it on both connection state and progress flows.